### PR TITLE
WT-1728 WT_SESSION.reset

### PR DESCRIPTION
**Pull request by daveh86**

### CODE REVIEW

File: src/session/session_api.c [hunk 1]
```diff
@@ -488,6 +488,57 @@ err:	API_END_RET_NOTFOUND_MAP(session, ret);
...
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+
+	conn = (WT_CONNECTION_IMPL *)wt_session->connection;
+	session = (WT_SESSION_IMPL *)wt_session;
+
+	SESSION_API_CALL(session, reset, config, cfg);
+
+	if (F_ISSET(&session->txn, WT_TXN_RUNNING))
+		WT_ERR_MSG(session, EINVAL, "transaction in progress");
```
Comment (keithbostic) [inline src/session/session_api.c]: Isn't reset one of those calls that doesn't fail? That is, after reset, the WT_SESSION holds no resources, even if the call "fails"?

Maybe reset could just abort the transaction.

**Comment by michaelcahill** [comment]
@daveh86, please fix the test failure: "RE-RUN dist/s_all and resubmit"

File: src/session/session_api.c (hunk 1 above)
Comment (michaelcahill) [inline src/session/session_api.c]: I don't want new calls to create transaction boundaries: that just makes application code hard to follow.  I think this check is reasonable: we could make reset succeed in this case, but it's not clear what it would mean.

The intention is that `WT_SESSION::reset` be called before a session is released into a cache, which would be a bad thing to do it a transaction were in progress.

**Comment by daveh86** [comment]
please retest this

**Comment by daveh86** [comment]
retest this please

**Comment by michaelcahill** [comment]
@agorrod, I think this is reasonable: can you please review?

File: src/session/session_api.c [hunk 2]
```diff
@@ -17,7 +17,7 @@ static int __session_rollback_transaction(WT_SESSION *, const char *);
  *	Reset all open cursors.
  */
 int
-__wt_session_reset_cursors(WT_SESSION_IMPL *session)
+__wt_session_reset_cursors(WT_SESSION_IMPL *session, int free_buffers )
```
Comment (agorrod) [inline src/session/session_api.c]: There is a trailing space after ```free_buffers```

File: src/include/wiredtiger.in [hunk 3]
```diff
@@ -1261,6 +1261,15 @@ struct __wt_session {
 	    const char *uri, const char *newuri, const char *config);
 
 	/*!
+	 * Reset the session. Any resources held by the session are released.
+	 * @snippet ex_all.c Reset the session
```
Comment (agorrod) [inline src/include/wiredtiger.in]: The above two lines aren't consistent with other comments. The form is more generally:

* One line with a very brief description, i.e: ```Reset the session.``` (I'd add ```handle``` to be more consistent with other APIs).
* A blank line, followed by the more detailed description of what the API does.
* Another blank line prior to the ```@snippet``` comment.

Comment (agorrod) [inline src/include/wiredtiger.in]: I also think it'd would be good to have more information, what resources are released, that an error will be generated if there is a transaction running, what happens if there are open cursors.

File: src/session/session_api.c [hunk 4]
```diff
@@ -488,6 +493,34 @@ err:	API_END_RET_NOTFOUND_MAP(session, ret);
...
+	SESSION_API_CALL_NOCONF(session, reset);
+
+	if (F_ISSET(&session->txn, WT_TXN_RUNNING))
+		WT_ERR_MSG(session, EINVAL, "transaction in progress");
+
+	WT_TRET(__wt_session_reset_cursors(session, 1));
+
+	WT_ASSERT(session, session->ncursors == 0);
+
+	/* Discard scratch buffers, error memory. */
```
Comment (agorrod) [inline src/session/session_api.c]: I don't think this comment adds value, and it's prone to become out of sync with the code.

File: src/session/session_api.c (hunk 1 above)
Comment (keithbostic) [inline src/session/session_api.c]: I agree, technically it's a transaction boundary, but so is `WT_CONNECTION.close`, for that matter. I wouldn't document applications can reasonably use `WT_SESSION.reset` to abort transactions, though.

Because it's an application error to call `WT_SESSION.reset` with a running transaction, I doubt if the application will be prepared to handle this error should it occur. Like you say, the application will call `WT_SESSION.reset`, log any errors, and put the handle into a cache.

In other words, if this ever happens, the application has already made a mistake, and if we don't kill the running transaction, it's likely the application has lost track of a transaction, which is worse?

Comment (michaelcahill) [inline src/session/session_api.c]: I think it's worse to allow this and have folks write bad application code.  This error will ensure that using WT_SESSION::reset is caught during development.  I'm much more comfortable with that than with having applications call WT_SESSION::reset when they actually want to roll back a running transaction.

Comment (agorrod) [inline src/session/session_api.c]: Hi,

I vote with Michael on this one, and it's consistent with what we do in
other APIs, including WT_SESSION::reconfigure and
WT_SESSION::transaction_sync

- Alex

**Comment by agorrod** [comment]
Thanks @daveh86 this lgtm now.

**Comment by agorrod** [comment]
@michaelcahill @keithbostic There is an outstanding question about the behavior of reset when there is a transaction running - can we resolve that after merging this?
